### PR TITLE
Migrate Guava cache usage to Caffeine

### PR DESCRIPTION
**Pull request by Stephan202**
For details, see #258. This is a draft PR to assess what is involved.

:exclamation: Note the base branch; these changes are built on top of #276 because Caffeine requires Java 8.

I'll add some comments with additional context/questions.

### CODE REVIEW

File: agent-bridge/src/main/java/com/newrelic/agent/bridge/ExtensionHolder.java [hunk 1]
```diff
@@ -17,7 +18,7 @@
      * Return the value associated with the instance in this ExtensionHolder. If the value is not present it will be
      * initialized (in a thread-safe manner) using valueLoader.
      */
-    public T getExtension(Object instance, Callable<T> valueLoader);
+    public T getExtension(Object instance, Supplier<T> valueLoader);
```
Comment (Stephan202) [inline agent-bridge/src/main/java/com/newrelic/agent/bridge/ExtensionHolder.java:21]: I found only one implementation of this interface. If it _is_ used elsewhere we should revert this change and introduce a `try/catch` construct in the implementation.

File: newrelic-agent/build.gradle [hunk 2]
```diff
@@ -94,6 +94,8 @@ dependencies {
 
     shadowIntoJar('com.google.guava:guava:28.2-jre') 
 
+    shadowIntoJar('com.github.ben-manes.caffeine:caffeine:2.9.0')
```
Comment (Stephan202) [inline newrelic-agent/build.gradle:97]: Caffeine 3.x requires Java 11.x; the 2.x line will [remain supported](https://github.com/ben-manes/caffeine/releases/tag/v3.0.0).

File: newrelic-agent/src/main/java/com/newrelic/agent/ThreadService.java [hunk 3]
```diff
@@ -82,20 +80,7 @@ public ThreadNameNormalizer getThreadNameNormalizer() {
...
-                    public String call() throws Exception {
-                        return threadInfo.getName();
-                    }
-                });
-            } catch (ExecutionException e) {
-                return threadInfo.getName();
-            }
-        }
-        return name;
+        return threadIdToName.get(threadInfo.getId(), id -> threadInfo.getName());
```
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/ThreadService.java:83]: Throughout the PR: `ExecutionException` isn't thrown anymore, so I dropped associated `try/catch` blocks where reasonable, taking into account whether the backing code could actually throw an exception and whether the fallback logic "made sense" or seemed to be there just to appease the compiler.

File: newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java [hunk 4]
```diff
@@ -37,14 +36,13 @@ public TimedTokenSet(int timeOut, TimeUnit unit, final ExpirationService expirat
         // onRemoval happens immediately after put, so we can hit onRemoval logic before getToken() even finishes
         long timeOutMilli = TimeConversion.convertToMilliWithLowerBound(timeOut, unit, 250L);
 
-        activeTokens = CacheBuilder.newBuilder()
-                .concurrencyLevel(8)
+        activeTokens = Caffeine.newBuilder()
+                .initialCapacity(8)
```
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java:40]: Here and further down: Caffeine [doesn't expose a `concurrencyLevel` setting](https://github.com/ben-manes/caffeine/blob/7f98672657a6e3515eac00ec86990b877a9d10dc/guava/src/test/java/com/google/common/cache/CacheBuilderFactory.java#L144-L150).  While working on this PR I found one source recommending to set the `initialCapacity` instead, but now I can't find it back.

CC @ben-manes in case this change makes no sense.

File: newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java [hunk 5]
```diff
@@ -37,14 +36,13 @@ public TimedTokenSet(int timeOut, TimeUnit unit, final ExpirationService expirat
...
                 .expireAfterAccess(timeOutMilli, TimeUnit.MILLISECONDS)
+                .executor(Runnable::run)
```
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java:42]: This `Runnable::run` is here because one of the tests expects to see the effects of the removal notification in a timely manner. See the [wiki](https://github.com/ben-manes/caffeine/wiki/Guava) on how Caffeine may schedule this kind of work on another thread. Alternatively the test could be tweaked. If the current behavior is desired, perhaps `.executor(Runnable::run)` should also be added in `AsyncTransactionService`.

File: newrelic-agent/src/main/java/com/newrelic/agent/extension/ExtensionRewriter.java [hunk 6]
```diff
@@ -27,7 +27,7 @@
 public class ExtensionRewriter {
 
     static final DependencyRemapper REMAPPER = new DependencyRemapper(ImmutableSet.of("org/objectweb/asm/",
-            "com/google/", "org/apache/commons/"));
+            "com/github/benmanes/caffeine/", "com/google/", "org/apache/commons/"));
```
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/extension/ExtensionRewriter.java:30]: I haven't investigated what this code actually does. Does it make sense to add Caffeine, or should it replace Guava, or...?

File: newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/weaver/ClassLoaderClassTransformer.java [hunk 7]
```diff
@@ -128,7 +128,7 @@ public ClassLoaderClassTransformer(InstrumentationProxy instrumentation, Set<Str
         // Try to use a custom Guava-based extension template to make NewField access better
         try {
             extensionTemplate = WeaveUtils.convertToClassNode(WeaveUtils.getClassBytesFromClassLoaderResource(
-                    GuavaBackedExtensionClass.class.getName(), GuavaBackedExtensionClass.class.getClassLoader()));
+                    CaffeineBackedExtensionClass.class.getName(), CaffeineBackedExtensionClass.class.getClassLoader()));
```
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/weaver/ClassLoaderClassTransformer.java:131]: `GuavaBackedExtensionClass` is now Caffeine-backend, so a rename seemed in order.

File: newrelic-agent/src/main/java/com/newrelic/agent/tracers/metricname/MetricNameFormats.java [hunk 8]
```diff
@@ -73,29 +60,11 @@ public static MetricNameFormat getFormatter(final Object invocationTarget, final
...
-                });
-            } catch (ExecutionException e) {
-                return new SimpleMetricNameFormat(getTracerMetricName(key.invocationTargetClassName, sig.getClassName(), metricName));
-            }
-        }
+        return cmsToMnf.get(
+                key,
+                k -> metricName == null
+                        ? sig.getMetricNameFormat(key.invocationTargetClassName, flags)
+                        : new SimpleMetricNameFormat(getTracerMetricName(key.invocationTargetClassName, sig.getClassName(), metricName)));
```
Comment (Stephan202) [inline newrelic-agent/src/main/java/com/newrelic/agent/tracers/metricname/MetricNameFormats.java:67]: Note that the `if` condition is now only evaluated on cache misses.

File: newrelic-agent/src/test/java/com/newrelic/agent/extension/DependencyRemapperTest.java [hunk 9]
```diff
@@ -72,8 +84,10 @@ public RewriteTest(int api) {
 
         @Override
         public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
-            // make sure we pick up this google reference
-            CacheBuilder<Object, Object> newBuilder = CacheBuilder.newBuilder();
+            // Make sure we pick up this Caffeine reference.
+            Caffeine<Object, Object> newBuilder = Caffeine.newBuilder();
+            // Make sure we pick up this Guava reference.
+            ImmutableSet<Object> set = ImmutableSet.of();
```
Comment (Stephan202) [inline newrelic-agent/src/test/java/com/newrelic/agent/extension/DependencyRemapperTest.java:90]: Related to my question above: no idea whether these test changes make sense.

File: newrelic-weaver/src/test/java/com/newrelic/weave/weavepackage/WeavePackageManagerTest.java [hunk 10]
```diff
@@ -239,7 +239,8 @@ public void testMaxInvalidClassLoaders() throws IOException {
             wpm.weave(cl, "com/newrelic/weave/weavepackage/WeavePackageManagerTest$NewNoMatchClass",
                     WeaveTestUtils.getClassBytes("com.newrelic.weave.weavepackage.WeavePackageManagerTest$NewNoMatchClass"));
         }
-        Assert.assertTrue(WeavePackageManager.MAX_INVALID_PACKAGE_CACHE >= wpm.invalidPackages.size());
+        wpm.invalidPackages.cleanUp();
```
Comment (Stephan202) [inline newrelic-weaver/src/test/java/com/newrelic/weave/weavepackage/WeavePackageManagerTest.java:242]: See the [wiki](https://github.com/ben-manes/caffeine/wiki/Guava); Caffeine's size-based eviction may cause the cache to grow _slightly_ larger than configured. Here we trigger an explicit cleanup to make sure the test passes. Under normal operation this kind of cleanup is performed in the context of subsequent cache operations.

**Review by Stephan202 — COMMENTED** [review]
Some context :)

File: newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java (hunk 4 above)
Comment (ben-manes) [reply newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java:40]: The FAQ entry on [Write Contention](https://github.com/ben-manes/caffeine/wiki/Faq#write-contention) does recommend increasing the `initialCapacity` if you want to speed up a warm-up phase. The number of locks is based on the hash table capacity (the hashbin is used as a lock), so resizes increases the lock count automatically. When the cache and many threads try to write to it, they will more likely collide on the same set of locks. The default is 16.

Comment (Stephan202) [reply newrelic-agent/src/main/java/com/newrelic/agent/TimedTokenSet.java:40]: Ah, that's where I read it indeed: 
> consider increasing the initial capacity to your expected maximum concurrency to compensate

Thanks as always :) I'll let the New Relic devs provide feedback on this, as I'm not familiar with this code's behavior at runtime.

**Comment by jasonjkeller** [comment]
Thanks for this draft @Stephan202 ! As soon as we have some bandwidth we'll take a closer look and see what we can do about getting this merged.

**Comment by Stephan202** [comment]
Hmm, deletion of the base PR also closed this PR. I can't edit the base branch either, I see. @jasonjkeller shall I open a new (non-draft) PR?

**Comment by jasonjkeller** [comment]
@Stephan202 Sorry about that! I didn't realize this occurred. The changes from `newrelic:build-with-java8` have been merged into `newrelic:main` I think if you wouldn't mind opening a new PR with your changes against `main` that would be the cleanest way to proceed. If it is still a work in progress then a draft PR is fine, otherwise go for a non-draft if it is ready to be reviewed. If you need any assistance getting the new PR open feel free to ping me. Thanks!

**Comment by Stephan202** [comment]
@jasonjkeller alright, I filed #295 and copied over the open comments. From my side no further changes are planned (pending review feedback, of course), so I filed it as a non-draft PR. Tnx for having a look once bandwidth is available :)
